store/index/gc.go

```diff
@@ -66,98 +65,197 @@ func (i *Index) garbageCollector(gcInterval time.Duration) {
+	index.bucketLk.Unlock()
+
+	if ctx.Err() != nil {
+		return 0, ctx.Err()
+	}
+
+	header, err := readHeader(index.headerPath)
+	if err != nil {
+		return 0, fmt.Errorf("cannot read index header: %w", err)
+	}
+
+	var rmCount int
+	basePath := index.basePath
+
+	index.flushLock.Lock()
+	lastFileNum := index.fileNum
+	index.flushLock.Unlock()
+
+	for fileNum := header.FirstFile; fileNum < lastFileNum; fileNum++ {
+		if _, busy := busySet[fileNum]; busy {
+			continue
 		}
-		if !stale {
-			break
+		indexPath := indexFileName(basePath, fileNum)
+
+		if fileNum == header.FirstFile {
+			header.FirstFile++
+			err = writeHeader(index.headerPath, header)
+			if err != nil {
+				return 0, fmt.Errorf("cannot write index header: %w", err)
+			}
+
+			err = os.Remove(indexPath)
+			if err != nil {
+				log.Errorw("Error removing index file", "err", err, "file", indexPath)
+				continue
+			}
+			log.Infow("Removed unused index file", "file", indexPath)
+			rmCount++
+			continue
 		}
-		fileNum++
-		header.FirstFile = fileNum
-		err = writeHeader(i.headerPath, header)
+
+		fi, err := os.Stat(indexPath)
 		if err != nil {
-			return 0, err
+			log.Errorw("Cannot stat index file", "err", err, "file", indexPath)
+			continue
+		}
+		if fi.Size() == 0 {
+			continue
 		}
-		// If updating index info ok, then remove stale index file.
-		err = os.Remove(indexPath)
+
+		err = os.Truncate(indexPath, 0)
 		if err != nil {
-			return 0, err
+			log.Errorw("Error truncating index file", "err", err, "file", indexPath)
+			continue
+		}
+		log.Infow("Emptied unused index file", "file", indexPath)
+
+		if ctx.Err() != nil {
+			break
 		}
-		count++
 	}
 
-	return count, nil
+	return rmCount, ctx.Err()
 }
 
 // gcIndexFile scans a single index file, checking if any of the entries are in
 // buckets that use this file. If no buckets are using this file for any of the
 // entries, then there are no more active entries and the file can be deleted.
-func (i *Index) gcIndexFile(ctx context.Context, fileNum uint32, indexPath string) (bool, error) {
-	file, err := openFileForScan(indexPath)
+func (index *Index) gcIndexFile(ctx context.Context, fileNum uint32, indexPath string) (bool, error) {
+	fi, err := os.Stat(indexPath)
+	if err != nil {
+		return false, fmt.Errorf("cannot stat index file: %w", err)
+	}
+	if fi.Size() == 0 {
+		// File is empty, so OK to delete if it is first file.
+		return true, nil
+	}
+
+	file, err := os.OpenFile(indexPath, os.O_RDWR, 0644)
 	if err != nil {
 		return false, err
 	}
 	defer file.Close()
 
-	inBuf := bufio.NewReader(file)
-	sizeBuffer := make([]byte, sizePrefixSize)
+	var freedCount int
+	var freeAtSize uint32
+	var busyAt, freeAt int64
+	freeAt = -1
+	busyAt = -1
+
+	sizeBuf := make([]byte, sizePrefixSize)
 	scratch := make([]byte, 256)
+	var pos int64
 	for {
 		if ctx.Err() != nil {
 			return false, ctx.Err()
 		}
-		if _, err = io.ReadFull(inBuf, sizeBuffer); err != nil {
+		if _, err = file.ReadAt(sizeBuf, pos); err != nil {
```

Comment:
should we still use bufio?

Reply:
Maybe. It depends on how many contiguous busy records there are.  If busy records are sparse, then no since it is better to skip over all the freed data instead of buffering it.

Reply:
After testing, the in-use index records become sparse very quickly.  Since free records get merged, what the index files look like is alternating free and busy records, where the free ones are very large and the busy ones are generally very small.  This means scanning can skip every other record, and the skips get large quickly.  So, not buffering is better in this case.